src/test/libraryfilesynctest.cpp

```diff
@@ -0,0 +1,712 @@
+#include <gtest/gtest.h>
+
+#include <QFile>
+#include <QFileInfo>
+#include <QTemporaryDir>
+#include <memory>
+
+#include "library/library_prefs.h"
+#include "sources/metadatasource.h"
+#include "test/librarytest.h"
+#include "track/track.h"
+#include "util/fileinfo.h"
+
+namespace {
+
+const auto kTestDir = QDir(QDir::current().absoluteFilePath("src/test/id3-test-data"));
+
+const auto kTestFileWithMetadata = QFileInfo(kTestDir, QStringLiteral("cover-test-jpg.mp3"));
+
+const auto kTestFileWithoutMetadata = QFileInfo(kTestDir, QStringLiteral("empty.mp3"));
+
+enum class FileType {
+    WithMetadata,
+    WithoutMetadata,
+};
+
+void sleepAfterFileLastModifiedUpdated() {
+    // Ensure that all subsequent modification time stamps will be
+    // strictly greater, i.e. strictly monotonic sequence of time
+    // stamps. This grace period seems to be mandatory for the CI
+    // test runs and might need to be adjusted depending on the
+    // results for different platforms.
+    QThread::msleep(1);
+}
+
+/// A temporary file system with a single audio file.
+class TempFileSystem {
+  public:
+    explicit TempFileSystem(const QFileInfo& testFile)
+            : m_fileInfo(m_tempDir.filePath(testFile.fileName())) {
+        createFile(testFile);
+    }
+
+    const mixxx::FileInfo& fileInfo() const {
+        return m_fileInfo;
+    }
+
+    QDateTime fileLastModified() const {
+        return m_fileInfo.toQFile().fileTime(QFileDevice::FileModificationTime);
+    }
+
+    void updateFileLastModified() const {
+        auto file = m_fileInfo.toQFile();
+        const auto oldLastModified = fileLastModified();
+        ASSERT_TRUE(oldLastModified.isValid());
+        // Rely on the OS to generate the time stamps instead of using
+        // QFileDevice::setFileTime() with a time stamp generated by
+        // QDateTime. The latter has proven to produce non-monotonic
+        // time stamps that cause spurious test failures.
+        ASSERT_TRUE(file.open(QIODevice::ReadWrite | QIODevice::ExistingOnly | QIODevice::Append));
+        const auto fileSize = file.size();
+        // Grow and shrink the file to bump the modification time stamp
+        ASSERT_TRUE(file.resize(fileSize + 1));
+        ASSERT_TRUE(file.flush());
+        ASSERT_TRUE(file.resize(fileSize));
+        file.close();
+        const auto newLastModified = fileLastModified();
+        ASSERT_TRUE(newLastModified.isValid());
+        ASSERT_GT(newLastModified, oldLastModified);
+        sleepAfterFileLastModifiedUpdated();
+    }
+
+    void removeFile() const {
+        ASSERT_TRUE(m_fileInfo.toQFile().remove());
+    }
+
+  private:
+    void createFile(const QFileInfo& testFile) {
+        ASSERT_TRUE(m_tempDir.isValid());
+        ASSERT_FALSE(m_fileInfo.exists());
+        mixxxtest::copyFile(
+                testFile.absoluteFilePath(),
+                m_fileInfo.location());
+        ASSERT_TRUE(m_fileInfo.exists());
+        sleepAfterFileLastModifiedUpdated();
```

Comment:
Please move that to the constructor or rename the function to createFileAndSleep() or createAndAgeFile() or such.

Reply:
Already tried that. Not possible due to the gtest macros.

Reply:
It looks like you answer belongs to a different thread.  

Here I have asked to move the sleep into the calling function or rename the function.

Reply:
No. The time machinery behind the scenes doesn't belong to the actual testing code. It is just there to ensure that all generated time stamps that are generated in quick succession during the tests are strictly monotonic.

This is the last time I am trying to explain it.